### PR TITLE
dan/per-10356-add-incremental-updates-to-the-factstore

### PR DESCRIPTION
**Pull request by danyi1212**

### CODE REVIEW

File: horizon/data_manager/policy_store.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                logger.error(f"Failed to insert fact: {res.status}\n{error}")
+            return res
+        except Exception as e:
+            logger.exception(f"Failed to insert fact: {e}")
+
+    async def delete_policy_data(
+        self, path: str = "", transaction_id: Optional[str] = None
+    ):
+        # TODO forward relevant objects to data manager instead of OPA
+        return super().delete_policy_data(path=path, transaction_id=transaction_id)
```
Comment (asafc) [inline horizon/data_manager/policy_store.py:139]: TODO: this requires implementation of **deletes** into the factstore.
it's ok (even recommended) to merge this now (to speed up testing) but @danyi1212 please open a ticket so we remember to close this gap later this week.

File: horizon/data_manager/policy_store.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    for target_type, target_objects in targets.items():
+                        for target in target_objects:
+                            # TODO missing subject_type and target_type
+                            await self._insert_fact(
+                                "relationships",
+                                {
+                                    "subject": subject,
+                                    "relation": relation,
+                                    "object": target,
+                                    "tenant": "",  # TODO unnecessary?
```
Comment (asafc) [inline horizon/data_manager/policy_store.py:63]: IMO opal data should definitely send the tenant information and we should sync it as part of the fact.
but not a blocker.

File: horizon/data_manager/policy_store.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                                },
+                            )
+            case ["role_assignments", full_user_key]:
+                user_key = full_user_key.lstrip("user:")
+                for subject, roles in policy_data.items():
+                    subject_type, subject_key = subject.split(":", 1)
+                    for role_key in roles:
+                        if subject_key == "__tenant":
```
Comment (asafc) [inline horizon/data_manager/policy_store.py:71]: did you mean:
```
if subject_type == "__tenant":
```
also __tenant should probably be a const

File: horizon/data_manager/policy_store.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                            )
+                        else:
+                            # TODO missing resource_type
+                            await self._insert_fact(
+                                "role_assignments",
+                                {
+                                    "actor": user_key,
+                                    "tenant": "",
+                                    "role": role_key,
+                                    "resource": subject_key,
```
Comment (asafc) [inline horizon/data_manager/policy_store.py:89]: IIRC the resource should be the entire subject - <id>:<type> like "document:budget"

**Review by asafc — APPROVED** [review]
overall looks ok - a few minor issues

File: horizon/data_manager/policy_store.py (hunk 2 above)
Comment (danyi1212) [reply horizon/data_manager/policy_store.py:63]: Relationship does not have a tenant, logically. The tenant association is for either of the resource instances at either end.

File: horizon/data_manager/update_operations.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    )
+                )
+            else:
+                yield InsertOperation(
+                    fact=Fact(
+                        type="role_assignments",
+                        attributes={
+                            "id": str(uuid4()),
+                            "actor": user_key,
+                            "tenant": "",
```
Comment (asafc) [inline horizon/data_manager/update_operations.py:57]: why drop the tenant? resource-role assignments have tenant.

File: horizon/data_manager/update_operations.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    user_key: str, data: JsonableValue
+) -> Iterator[AnyOperation]:
+    attributes = data.get("attributes", {})
+    yield InsertOperation(
+        fact=Fact(
+            type="users",
+            attributes={
+                "id": user_key,
+                "attributes": json.dumps(attributes),
+                # TODO remove the json.dumps after fixing the map[string]string issue in Go
```
Comment (asafc) [inline horizon/data_manager/update_operations.py:75]: nonblocker: don't like that we are simply passing through attributes but we'll merge first and close the schema format + validation later

File: horizon/data_manager/update_operations.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    instance_key: str, data: JsonableValue
+) -> Iterator[AnyOperation]:
+    attributes = data.get("attributes", {})
+    yield InsertOperation(
+        fact=Fact(
+            type="resource_instance",
+            attributes={
+                "id": instance_key,
+                "attributes": json.dumps(attributes),
+                # TODO remove the json.dumps after fixing the map[string]string issue in Go
```
Comment (asafc) [inline horizon/data_manager/update_operations.py:91]: same here

nonblocker: don't like that we are simply passing through attributes but we'll merge first and close the schema format + validation later

**Review by asafc — APPROVED** [review]
approved with comment (1 blocker)

**Comment by danyi1212** [comment]
Still missing the delete operations

File: horizon/data_manager/update_operations.py [hunk 8]
```diff
@@ -58,37 +75,77 @@ def _get_operations_for_update_role_assigment(
...
+            fact=Fact(
+                type="relationship_tuples",
+                attributes={"subject": instance_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="role_assignments",
+                attributes={"resource": instance_key},
+            ),
```
Comment (asafc) [inline horizon/data_manager/update_operations.py:150]: why? a resource instance can exist without attributes? why remove it?

File: horizon/data_manager/update_operations.py [hunk 9]
```diff
@@ -58,37 +75,77 @@ def _get_operations_for_update_role_assigment(
...
+            fact=Fact(
+                type="users",
+                attributes={"id": user_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="role_assignments",
+                attributes={"actor": user_key},
+            ),
```
Comment (asafc) [inline horizon/data_manager/update_operations.py:108]: same question here?

**Review by asafc — COMMENTED** [review]
it's still ok - but you did not fix my previous blocking comment about tenant in resource role assignment and i am not sure about the logic of "update user" and "update instance" becomes delete. is this because we are using PUT? if so? where are the comments that explain it?

File: horizon/data_manager/update_operations.py (hunk 8 above)
Comment (danyi1212) [reply horizon/data_manager/update_operations.py:150]: When we "delete" an object from OPA we simply send it without any attributes, that's how the data update used to work. The attributes are not just the custom attributes, but contains the key as well, so if it doesn't exist then it's a delete operation

File: horizon/data_manager/update_operations.py (hunk 5 above)
Comment (danyi1212) [reply horizon/data_manager/update_operations.py:74]: resource role assignment does not have tenant in the data object. In tenant role assignments, the tenant key is passed as the subject, where it's like `__tenant:<key>`, but on resource role assignments it's like `<resource>:<key>`

File: horizon/data_manager/policy_store.py (hunk 2 above)
Comment (asafc) [reply horizon/data_manager/policy_store.py]: i disagree but we can sort that out later when we decide on the schema.

File: horizon/data_manager/update_operations.py (hunk 5 above)
Comment (asafc) [reply horizon/data_manager/update_operations.py:74]: i see - this is a bug in OPAL data generator, i will approve for now, we'll need to fix it.
i'll leave it to the schema discussion.

File: horizon/data_manager/update_operations.py (hunk 8 above)
Comment (asafc) [reply horizon/data_manager/update_operations.py:150]: ok, please add that in a comment.

File: horizon/data_manager/update_operations.py (hunk 9 above)
Comment (asafc) [reply horizon/data_manager/update_operations.py:108]: please add comment

**Review by asafc — APPROVED** [review]
please add comments where i requested

**Pull request merged**